File: javascript/ast.js
'use strict';

var ADD       = 0;
var SUBSTRACT = 1;
var MULTIPLY  = 2;
var DEVIDE    = 3;
var SUM       = 4;

var functions = [];
functions[ADD]       = function(args) { return +args[0] + +args[1]; };
functions[SUBSTRACT] = function(args) { return +args[0] - +args[1]; };
functions[MULTIPLY]  = function(args) { return +args[0] * +args[1]; };
functions[DEVIDE]    = function(args) { return +args[0] / +args[1]; };

functions[SUM] = function(args) {
    var sum = 0;
    for ( var i = 0; i < args.length; i++ ) {
        sum += +args[i];
    }
    return sum;
};


function evaluateAST(ast) {
    var oper = ast[0];
    var func = functions[oper];

    var evaluatedArgs = [];

    for (var i = 1; i < ast.length; i++) {
        if ( Array.isArray(ast[i]) )  {
            evaluatedArgs.push( evaluateAST(ast[i]) );
        } else {
            evaluatedArgs.push( ast[i] );
        }
    }

    return func(evaluatedArgs);
}

function timeAST(ast) {
    var sum = 0;
    var iterations = 100000;

    console.time('COMPUTE AST');

    for (var i = 0; i < iterations; i++) {
        sum += evaluateAST(ast);
    }

    console.timeEnd('COMPUTE AST');

    if ( Math.abs(sum - 3900000) > 0.001 ) throw 'WRONG SUM ' + sum;
}

var ast = [ SUM,
    [ SUBSTRACT,[ADD,[DEVIDE,[MULTIPLY,10,20],30],40],50],
    [ SUBSTRACT,[ADD,[DEVIDE,[MULTIPLY,20,30],40],50],60],
    [ SUBSTRACT,[ADD,[DEVIDE,[MULTIPLY,30,40],50],60],70],
    [ SUBSTRACT,[ADD,[DEVIDE,[MULTIPLY,40,50],60],70],80]
];


timeAST(ast);

File: go/ast.go
package main

import (
	"fmt"
	"math"
	"time"
)

/*

This is NOT idiomatic Go code. You should never write Go code like this.

*/

const (
	ADD = iota
	SUBSTRACT
	MULTIPLY
	DEVIDE
	SUM
)

var functions = map[int]func(args []float64) float64{
	ADD: func(args []float64) float64 {
		return args[0] + args[1]
	},

	SUBSTRACT: func(args []float64) float64 {
		return args[0] - args[1]
	},

	MULTIPLY: func(args []float64) float64 {
		return args[0] * args[1]
	},

	DEVIDE: func(args []float64) float64 {
		return args[0] / args[1]
	},

	SUM: func(args []float64) float64 {
		sum := 0.0

		for _, arg := range args {
			sum += arg
		}

		return sum
	},
}

func evaluateAST(ast []interface{}) float64 {
	oper := ast[0].(int)
	funct := functions[oper]

	evaluatedArgs := make([]float64, 0, len(ast)-1) // 1. cuts 30%

	for _, v := range ast[1:] { // 2. cuts 10%
		if e, ok := v.([]interface{}); ok {
			evaluatedArgs = append(evaluatedArgs, evaluateAST(e))
		} else {
			evaluatedArgs = append(evaluatedArgs, float64(v.(int)))
		}
	}

	return funct(evaluatedArgs)
}

func timeAST(ast []interface{}) {
	iterations := 100000
	sum := 0.0

	time_start := time.Now()

	for i := 0; i < iterations; i++ {
		sum += evaluateAST(ast)
	}

	compute_time := time.Since(time_start)

	fmt.Printf("COMPUTED [%d] ITERATIONS IN [%f] SECONDS\n", iterations, compute_time.Seconds())

	if math.Abs(sum-3900000.0) > 0.001 {
		panic(fmt.Sprintf("WRONG SUM %f", sum))
	}
}

func main() {
	ast := []interface{}{SUM,
		[]interface{}{SUBSTRACT, []interface{}{ADD, []interface{}{DEVIDE, []interface{}{MULTIPLY, 10, 20}, 30}, 40}, 50},
		[]interface{}{SUBSTRACT, []interface{}{ADD, []interface{}{DEVIDE, []interface{}{MULTIPLY, 20, 30}, 40}, 50}, 60},
		[]interface{}{SUBSTRACT, []interface{}{ADD, []interface{}{DEVIDE, []interface{}{MULTIPLY, 30, 40}, 50}, 60}, 70},
		[]interface{}{SUBSTRACT, []interface{}{ADD, []interface{}{DEVIDE, []interface{}{MULTIPLY, 40, 50}, 60}, 70}, 80},
	}

	timeAST(ast)
}
